pins/_adaptors.py

```diff
@@ -0,0 +1,172 @@
+from __future__ import annotations
+
+import json
+from abc import abstractmethod
+from typing import TYPE_CHECKING, Any, ClassVar, TypeAlias, overload
+
+from typing_extensions import Self
```

Comment:
I'll need to avoid using `Self`. I forget that `typing_extensions` isn't a built-in...